ir/src/tests/mod.rs

```diff
@@ -379,6 +379,120 @@ fn transition_constraints_using_parens() {
     assert!(result.is_ok());
 }
 
+#[test]
+fn random_values_fixed_list() {
+    let source = "
+    trace_columns:
+        main: [a, b[12]]
+    public_inputs:
+        stack_inputs: [16]
+    random_values:
+        rand: [16]
+    integrity_constraints:
+        enf a' = $rand[3] + 1
+    boundary_constraints:
+        enf a.first = $rand[10] * 2
+        enf a.last = 1";
+
+    let parsed = parse(source).expect("Parsing failed");
+    let result = AirIR::from_source(&parsed);
+    assert!(result.is_ok());
+}
+
+#[test]
+fn random_values_custom_name() {
+    let source = "
+    trace_columns:
+        main: [a, b[12]]
+    public_inputs:
+        stack_inputs: [16]
+    random_values:
+        alphas: [16]
+    integrity_constraints:
+        enf a' = $alphas[3] + 1
+    boundary_constraints:
+        enf a.first = $alphas[10] * 2
+        enf a.last = 1";
+
+    let parsed = parse(source).expect("Parsing failed");
+    let result = AirIR::from_source(&parsed);
+    assert!(result.is_ok());
+}
+
+#[test]
+fn random_values_ident_vector() {
+    let source = "
+    trace_columns:
+        main: [a, b[12]]
+    public_inputs:
+        stack_inputs: [16]
+    random_values:
+        rand: [c, d[4]]
+    integrity_constraints:
+        enf a' = c + d[2] + $rand[1]
+    boundary_constraints:
+        enf a.first = (d[1] - $rand[0]) * 2
+        enf a.last = c";
+
+    let parsed = parse(source).expect("Parsing failed");
+    let result = AirIR::from_source(&parsed);
+    assert!(result.is_ok());
+}
+
+#[test]
+fn error_random_values_out_of_bounds() {
+    let source_fixed_list = "
+    trace_columns:
+        main: [a, b[12]]
+    public_inputs:
+        stack_inputs: [16]
+    random_values:
+        rand: [4]
+    integrity_constraints:
+        enf a' = $rand[4] + 1
+    boundary_constraints:
+        enf a.first = $rand[10] * 2
+        enf a.last = 1";
+
+    let parsed = parse(source_fixed_list).expect("Parsing failed");
+    let result = AirIR::from_source(&parsed);
+    assert!(result.is_err());
+
+    let source_ident_vector_sub_vec = "
+    trace_columns:
+        main: [a, b[12]]
+    public_inputs:
+        stack_inputs: [16]
+    random_values:
+        rand: [c, d[4]]
+    integrity_constraints:
+        enf a' = c + 1
+    boundary_constraints:
+        enf a.first = d[5] * 2
+        enf a.last = 1";
+
+    let parsed = parse(source_ident_vector_sub_vec).expect("Parsing failed");
+    let result = AirIR::from_source(&parsed);
+    assert!(result.is_err());
+
+    let source_ident_vector_index = "
+    trace_columns:
+        main: [a, b[12]]
+    public_inputs:
+        stack_inputs: [16]
+    random_values:
+        rand: [c, d[4]]
+    integrity_constraints:
+        enf a' = c + 1
+    boundary_constraints:
+        enf a.first = $rand[10] * 2
+        enf a.last = 1";
+
+    let parsed = parse(source_ident_vector_index).expect("Parsing failed");
+    let result = AirIR::from_source(&parsed);
+    assert!(result.is_err());
+}
+
```

Comment:
I think we should change these so that random values are only used against auxiliary columns. We don't have a check for it yet but we should still have valid tests here for now. Also, do you think we should address #46 here as well or should we do that in another PR?

Reply:
I think I can do it in this PR, looks like it wouldn't involve a lot of changes. Just one question: should we also throw an error if random values are used against the main trace in integrity constraints?

Reply:
I don't think so. I think we can have integrity constraints with random values and main trace column access but maybe it should be an error if there are random values but no auxiliary column access in a constraint? I'm not sure, maybe @grjte or @bobbinth could confirm?

Reply:
Ideally, something like this should be a warning - but since we don't have warnings, I'm not quite sure what's the best approach. Probably having it be an error is better - but if it creates a lot of extra work now, we can do it later too.